lib/ra10ke/diff.rb

```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+require 'git'
+require 'r10k/puppetfile'
+
+module Ra10ke
+  module Diff
+    class Task
+      attr_reader :repo, :puppetfile
+
+      def initialize
+        @repo = Git.open('.')
+        @original_branch = repo.current_branch
+      end
+
+      def run(branch_a:, branch_b:)
+        validate_clean_repo
+        validate_branches(branch_a, branch_b)
+
+        modules_a = load_modules(branch_a)
+        modules_b = load_modules(branch_b)
+
+        differences = calculate_differences(modules_a, modules_b)
+        print_differences(differences)
+      ensure
+        restore_original_branch
+      end
+
+      private
+
+      def validate_clean_repo
+        return if repo.status.changed.empty? && repo.status.untracked.empty?
+
+        puts 'Git repository is not clean. Please commit or stash your changes before running this task.'
```

Comment:
There is also the warn method that will write a message to stderr and then does exit 1.

Reply:
okay, thx for the hint, but why not fail?

Reply:
can you pls go into more detail? 

this does not exit 1:
```
warn("some text")
```

Reply:
or are you refering to `abort("test")`?

Reply:
as discussed in chat: we go with abort, because the other task also do that. if this is not okay somehow anymore we do a another change to update to warn or whatever is now the best solution to do this.